File: bindings/cpp/tests/inproc_context.cpp
#include "harness.hpp"

#include <atomic>
#include <chrono>
#include <string>
#include <thread>
#include <utility>
#include <vector>

namespace {

std::string bytes_to_string(const std::vector<uint8_t> &body) {
  return std::string(reinterpret_cast<const char *>(body.data()), body.size());
}

std::string bytes_to_string(robot_bus::BytesView payload) {
  return std::string(reinterpret_cast<const char *>(payload.data), payload.size);
}

}  // namespace

int main() {
  using namespace robot_bus;
  using namespace robot_bus::test;

  Context ctx;
  RobotBusBrokerOptions opts = ephemeral_tcp_opts(0, 1);  // keep inproc (+ ipc)
  Broker broker(ctx, opts);
  std::this_thread::sleep_for(std::chrono::milliseconds(150));

  {
    std::atomic<int> hits{0};
    auto sub = Node::inproc_with_context(ctx, "inproc-sub");
    auto sub_h = sub.create_subscription("/inproc/demo", [&](std::string_view, BytesView payload) {
      ROBOT_BUS_CHECK(bytes_to_string(payload) == "hello-inproc");
      hits.fetch_add(1);
    });
    sub.start();
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    auto pub = Node::inproc_with_context(ctx, "inproc-pub");
    auto topic = pub.create_publisher("/inproc/demo");
    const char *msg = "hello-inproc";
    topic.publish(BytesView(reinterpret_cast<const uint8_t *>(msg), 12));

    auto deadline = std::chrono::steady_clock::now() + std::chrono::seconds(3);
    while (hits.load() < 1 && std::chrono::steady_clock::now() < deadline) {
      std::this_thread::sleep_for(std::chrono::milliseconds(20));
    }
    ROBOT_BUS_CHECK(hits.load() >= 1);

    sub.stop();
    (void)sub_h;
  }

  {
    auto server = Node::inproc_with_context(ctx, "inproc-action-server");
    auto act = server.create_action_server("/inproc/action", [](BytesView body) {
      const std::string payload = bytes_to_string(body);
      const std::string fb = "step:" + payload;
      const std::string res = "done:" + payload;
      std::vector<std::pair<std::string, std::vector<uint8_t>>> phases;
      phases.emplace_back("FEEDBACK", std::vector<uint8_t>(fb.begin(), fb.end()));
      phases.emplace_back("RESULT", std::vector<uint8_t>(res.begin(), res.end()));
      return phases;
    });
    server.start();
    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    auto client_node = Node::inproc_with_context(ctx, "inproc-action-client");
    auto client = client_node.create_action_client("/inproc/action");
    std::vector<std::string> feedback;
    const std::string goal_body = "move";
    auto handle = client.send_goal(
        goal_body,
        [&](const ActionMessage &message) { feedback.push_back(bytes_to_string(message.body)); },
        nullptr, 3.0);

    ROBOT_BUS_CHECK(handle.action_name() == "/inproc/action");
    ROBOT_BUS_CHECK(!handle.goal_id().empty());
    auto result_message = handle.wait_result(3.0);
    ROBOT_BUS_CHECK(result_message.kind == "RESULT");
    ROBOT_BUS_CHECK(bytes_to_string(result_message.body) == "done:move");
    ROBOT_BUS_CHECK(feedback.size() == 1);
    ROBOT_BUS_CHECK(feedback[0] == "step:move");

    server.shutdown();
    server.wait();
    (void)act;
  }

  broker.stop();
  return 0;
}

File: tests/inproc_context.rs
//! Same-process inproc requires a shared [`robot_bus::Context`].

mod support;

use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

#[cfg(feature = "console")]
use robot_bus::ConsoleBrokerConfig;
use robot_bus::{
    ActionKind, ActionMessage, Context, HighWaterMark, Node, NodeOptions, Publisher,
    RobotBusBroker, RobotBusConfig,
};
use support::lock_brokers;

fn inproc_broker_config() -> RobotBusConfig {
    // Ephemeral TCP/gRPC ports — defaults (e.g. :15570) collide under parallel cargo test.
    let mut config = support::ephemeral_robot_bus_config();
    // Keep inproc (+ ipc/tcp) so Node::inproc_with_context can reach the proxy.
    config.message.bind_all_transports = true;
    config.service.bind_all_transports = true;
    config.action.bind_all_transports = true;
    #[cfg(feature = "console")]
    {
        config.console = ConsoleBrokerConfig {
            enabled: false,
            ..ConsoleBrokerConfig::default()
        };
    }
    config
}

#[test]
fn inproc_pubsub_with_shared_context() {
    let _guard = lock_brokers();
    let ctx = Context::new();
    let broker =
        RobotBusBroker::start_with_context(&ctx, inproc_broker_config()).expect("broker");
    thread::sleep(Duration::from_millis(150));

    let hits = Arc::new(AtomicUsize::new(0));
    let hits_cb = Arc::clone(&hits);

    let mut sub = Node::inproc_with_context(&ctx, "inproc-sub");
    sub.set_stream_hwm(HighWaterMark {
        snd: 1000,
        rcv: 1000,
    })
    .expect("set_stream_hwm");
    sub.create_subscription_raw(
        "/inproc/demo",
        Arc::new(move |_topic, payload| {
            assert_eq!(payload, b"hello-inproc");
            hits_cb.fetch_add(1, Ordering::SeqCst);
        }),
        None,
    )
    .expect("subscribe");

    let shutdown = sub.shutdown_handle().expect("shutdown handle");
    let xsub = NodeOptions::inproc().message_xsub_endpoint().expect("xsub");
    let pub_ctx = ctx.clone();
    let hits_wait = Arc::clone(&hits);
    let worker = thread::spawn(move || {
        thread::sleep(Duration::from_millis(200));
        let publisher = Publisher::with_shared_context(
            &pub_ctx,
            Some(&xsub),
            HighWaterMark {
                snd: 1000,
                rcv: 1000,
            },
        )
        .expect("publisher");
        let deadline = Instant::now() + Duration::from_secs(5);
        while hits_wait.load(Ordering::SeqCst) == 0 {
            if Instant::now() >= deadline {
                break;
            }
            let _ = publisher.publish("/inproc/demo", b"hello-inproc");
            thread::sleep(Duration::from_millis(20));
        }
        shutdown.shutdown();
    });

    let _ = sub.spin();
    worker.join().expect("publisher thread");
    assert!(
        hits.load(Ordering::SeqCst) > 0,
        "expected at least one inproc message"
    );

    broker.stop().expect("stop broker");
}

#[test]
fn inproc_action_goal_handle() {
    let _guard = lock_brokers();
    let ctx = Context::new();
    let broker =
        RobotBusBroker::start_with_context(&ctx, inproc_broker_config()).expect("broker");
    thread::sleep(Duration::from_millis(150));

    let mut server = Node::inproc_with_context(&ctx, "inproc-action-server");
    server
        .create_action_server_raw(
            "/inproc/action",
            Arc::new(|body| {
                vec![
                    ("FEEDBACK".into(), [b"step:", body].concat()),
                    ("RESULT".into(), [b"done:", body].concat()),
                ]
            }),
            None,
        )
        .expect("create_action_server_raw");
    server.start().expect("start");
    thread::sleep(Duration::from_millis(100));

    let mut client_node = Node::inproc_with_context(&ctx, "inproc-action-client");
    let client = client_node
        .create_action_client_raw("/inproc/action")
        .expect("create_action_client_raw");
    let feedbacks = Arc::new(Mutex::new(Vec::new()));
    let callback_feedbacks = Arc::clone(&feedbacks);
    let goal = client
        .send_goal(
            b"move",
            None,
            Some(Duration::from_secs(3)),
            Some(Arc::new(move |message: &ActionMessage| {
                callback_feedbacks
                    .lock()
                    .expect("feedback mutex")
                    .push(message.body.clone());
            })),
        )
        .expect("send_goal");
    assert_eq!(goal.action_name(), "/inproc/action");
    assert!(!goal.goal_id().is_empty());
    let result = goal.wait_result().expect("wait_result");
    assert_eq!(result.kind, ActionKind::Result);
    assert_eq!(result.body, b"done:move");
    assert_eq!(
        *feedbacks.lock().expect("feedback mutex"),
        vec![b"step:move".to_vec()]
    );

    server.shutdown().expect("shutdown");
    server.stop().expect("stop");
    server.wait().expect("wait");
    broker.stop().expect("stop broker");
}
